rfcs/0147-third-party-login.md

```diff
@@ -0,0 +1,230 @@
+# RFC 147 - Third-Party Login
+* Comments: [#147](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/147)
+* Proposed by: @djmitche
+
+# Background
+
+Taskcluster is an API service and has no notion of “users”.
+Instead, it respects Taskcluster credentials which carry authorization information with them.
+
+However, Taskcluster has a browser-based user interface, and users can “sign in” to that interface using one of a variety of authentication providers such as Auth0 or GitHub.
+When they do so, they are given credentials, and the authorization information in those credentials are derived from details of their user account such as group membership.
+
+Other, third-party (non-Taskcluster) services also accept Taskcluster credentials -- this is especially common for CI-related services.
+The interfaces for such services typically need to do their own user authentication, but also want to utilize Taskcluster credentials.
+
+We also have a command-line tool which can interact with a local browser to fetch credentials and place them in environment variables for use by scripts.
+This tool’s functionality is similar to `heroku login`, for example.
+
+## History
+
+### Early Revisions
+
+Long ago, we supported this via a simple redirect flow, where a site redirected the user to `https://login.taskcluster.net/` with a callback URL and some other information.
+The login site gathered user consent and redirected the user back with taskcluster credentials in the URL.
+This supported both third-party login with other sites and command-line login.
+Problems with this approach included
+
+* The redirect flow and unfamiliar appearance of the site caused user confusion
+* Credentials were (arbitrarily) limited to 3 days, so users had to re-consent every 3 days (note that we did “whitelist” some specific third parties, keyed by callback URL)
+* Many federated sites wanted user information, and Taskcluster does not provide that
+
+The returned credentials always contain the full authorization for the user, even if only a small subset are required.
+
+### Current Implementation
```

Comment:
SPAs using the implicit flow are a normal login process. Note that for the most part where this doc says "auth0" you can replace by OAuth2.

The part that is specific to auth0 (really, Mozilla) is how group access is transmitted in a claim through the user info endpoint, id token and in auth0 and CIS APIs.

Reply:
Thanks for peeking in!

It's not that we have concerns with Auth0.  It's that we do not want to require our users to use Auth0 + People API.  I'm not sure what you mean by either of the two bulleted issues.  We discussed that in our meeting.  If there's further concerns or confusion over the content of this RFC, I'd be happy to talk about it again.

Reply:
> we do not want to require our users to use Auth0 + People API

Why? SSO is the company standard, actively managed by a well staffed team, and already on the critical path of our services. I imagine you have strong reasons for deciding not to use it and, instead, pay the very high price of re-implementing an identity provider on your own. Would it be possible to list those reasons here?

Reply:
Ah, you were not at the meeting I mentioned.  Our users are not necessarily at Mozilla.

Reply:
..and a key clarification we found in Zoom: by "users" I mean "people deploying Taskcluster", not Firefox developers.

Reply:
@djmitche just to clarify the part that you might find most useful right now: we should be able to skip the consent dialog if you want us to do so (f.e. if you have 2 consent dialogs when using SSO: auth0's and then your own, but want to hide the auth0/SSO's one)

Reply:
Thanks! There's only one consent dialog right now (auth0's), so I don't think there's any need to change.

I think the sentence about consent you're referring to is:

> Auth0’s implementation of “consent” has not developed as expected

and that's mostly about the text that it shows in the dialog.  If I remember (I consented months ago so I can't see), it's something like "This application wants access to your auth tenant: taskcluster-credentials".  There's not much capacity to modify the verbiage that appears there to say something clearer.

But even if that were fixed, I think we would still want to proceed with the proposed changes.